### PR TITLE
feat: log accuracy testing to testing server

### PR DESCRIPTION
**Pull request by anaprietonem**
## Description
Extend [test_accuracy](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) to also record a scalar per commit on the integration-test server, alongside the existing MLflow comparison.

- New [track_accuracy_results(test_case, final_loss)](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) in [benchmark_server.py](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html), mirroring [track_dataloader_benchmark_results](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) (main-only, rank-zero).
- New [get_benchmark_store(kind)](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) helper: yaml [path](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) is a base dir, [<kind>](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) is appended ([benchmarks/](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html), accuracy/); optional [paths:](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) mapping overrides per kind.
- [test_accuracy.py](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) pushes the last train_multi_dataset_loss_epoch value under accuracy/global/accuracyFinalLoss.
- [test_benchmark.py](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) switched to [get_benchmark_store("benchmarks")](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html).
- Docs updated in [benchmarking.rst](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [contributing.rst](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html).

Config note: if your [~/.config/anemoi/anemoi-benchmark.yaml](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) [path](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) currently ends in [/benchmarks](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html), shorten it to the parent (or add a [paths:](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) block) — otherwise benchmarks land in [.../benchmarks/benchmarks/](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html).

## What problem does this change solve?
[](https://github.com/ecmwf/anemoi-core/issues/1255)

## What issue or task does this change relate to?
[](https://github.com/ecmwf/anemoi-core/issues/1255)

##  Additional notes ##

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/Contributor-License-Agreement.md)

----
📚 Documentation preview 📚: https://anemoi-training--1256.org.readthedocs.build/en/1256/

----
📚 Documentation preview 📚: https://anemoi-graphs--1256.org.readthedocs.build/en/1256/

----
📚 Documentation preview 📚: https://anemoi-models--1256.org.readthedocs.build/en/1256/

### CODE REVIEW

**Comment by anaprietonem** [comment]
Before merging this we'd need to update the path ~/.config/anemoi/anemoi-benchmark.yaml either !!!

File: training/src/anemoi/training/diagnostics/benchmark_server.py [hunk 1]
```diff
@@ -49,6 +49,30 @@ def parse_benchmark_config(path: Path) -> tuple[str, str, str]:
     return user, hostname, path
 
 
+DEFAULT_BENCHMARK_CONFIG_PATH = Path("~/.config/anemoi/anemoi-benchmark.yaml").expanduser()
```
Comment (mcocdawc) [inline training/src/anemoi/training/diagnostics/benchmark_server.py]: could be declared `Final`

File: training/src/anemoi/training/diagnostics/benchmark_server.py [hunk 2]
```diff
@@ -720,6 +742,42 @@ def track_dataloader_benchmark_results(
         benchmark_server.set_value(dataloader_benchmark_value)
 
 
+@rank_zero_only
+def track_accuracy_results(
```
Comment (mcocdawc) [inline training/src/anemoi/training/diagnostics/benchmark_server.py]: Would it be possible to accept a key and a scalar, such that we can create specialiced functions for different metrices; something along the lines of:

```python
from functools import partial

track_accuracy_final_loss = partial(track_accuracy_result, name="accuracyFinalLoss")

...
track_accuracy_final_loss(float(epoch_history[-1].value))
```

I have the feeling that we want to play around with different metrices in the future?

**Review by mcocdawc — COMMENTED** [review]
A small generalization might make it possible to seamlessly track different  accuracy metrices in the future. I have the feeling that this is something people will want to use very soon.

File: training/tests/integration/test_accuracy.py [hunk 3]
```diff
@@ -104,3 +106,26 @@ def assert_similar(run_id1: str, run_id2: str) -> None:
...
+        msg = f"Run {trainer.run_id} has no '{epoch_metric}' history on {mlflow_server}"
+        raise ValueError(msg)
+    final_loss = float(epoch_history[-1].value)
+    track_accuracy_results(test_case="global", final_loss=final_loss)
+
+    prune_mlflow_runs(
+        client=client,
+        experiment_name=config.diagnostics.log.mlflow.experiment_name,
+        run_name=config.diagnostics.log.mlflow.run_name,
+        protected_run_ids=[reference_id],
```
Comment (mcocdawc) [inline training/tests/integration/test_accuracy.py]: I would like to point out a race condition here.
If there are two feature branches which both change the reference id and they are repeatedly run, they might exceed the last n runs and start deleting each others references.

This will show up as a loud explicit failure.

Preventing this makes the code much more complicated, so I would leave it as is, perhaps increase n and only if it turns out to be a problem fix it.

File: training/src/anemoi/training/diagnostics/benchmark_server.py (hunk 2 above)
Comment (anaprietonem) [reply training/src/anemoi/training/diagnostics/benchmark_server.py]: Neat, yes good idea, I have added this now!

File: training/docs/user-guide/benchmarking.rst [hunk 4]
```diff
@@ -859,7 +859,41 @@ The server location is read from a file
...
+kinds of integration tests: results land under ``<path>/<kind>``, where
+``kind`` is chosen by each test (currently ``benchmarks`` for the
+performance benchmarks in ``test_benchmark.py`` and ``accuracy`` for the
+loss-curve regression check in ``test_accuracy.py``). For example, with
+
+.. code:: yaml
+
+   user: data
+   hostname: ecmwf.int
+   path: /home/data/public/anemoi-integration-tests/training
```
Comment (cathalobrien) [inline training/docs/user-guide/benchmarking.rst]: can you replace this with a fake path please
```
/path/to/benchmark/results
```

**Review by cathalobrien — COMMENTED** [review]
looks good to me :)

File: training/docs/user-guide/benchmarking.rst (hunk 4 above)
Comment (anaprietonem) [reply training/docs/user-guide/benchmarking.rst]: good catch, I updated it now!

**Review by cathalobrien — APPROVED** [review]
LGTM pending integration tests passing. nice work!
